src/main/scala/exu/dispatch.scala

```diff
@@ -160,21 +172,49 @@ class SliceDispatcher(implicit p: Parameters) extends Dispatcher
   assert(!(b_valid && b_busy_resp.prs1_busy && b_head.lrs1 === 0.U), "[rename] x0 is busy??")
   assert(!(b_valid && b_busy_resp.prs2_busy && b_head.lrs2 === 0.U), "[rename] x0 is busy??")
 
+  if(usingFPU){
+    io.slice_fp_busy_req_uops.get(0) := a_head
+    io.slice_fp_busy_req_uops.get(1) := b_head // also needed for b for float loads
+    val a_flt_busy_resp = io.slice_fp_busy_resps.get(0)
+    val b_flt_busy_resp = io.slice_fp_busy_resps.get(1)
+    // fp busy info
+    when(a_head.lrs1_rtype === RT_FLT){
+      a_head.prs1_busy := a_flt_busy_resp.prs1_busy
+    }
+    when(a_head.lrs2_rtype === RT_FLT){
+      a_head.prs2_busy := a_flt_busy_resp.prs2_busy
+    }
+    a_head.prs3_busy := a_head.frs3_en && a_flt_busy_resp.prs3_busy
+    when(b_head.lrs1_rtype === RT_FLT){
+      b_head.prs1_busy := b_flt_busy_resp.prs1_busy
+    }
+    when(b_head.lrs2_rtype === RT_FLT){
+      b_head.prs2_busy := b_flt_busy_resp.prs2_busy
+    }
+    b_head.prs3_busy := b_head.frs3_en && b_flt_busy_resp.prs3_busy
+  }
+
   // this is handling the ready valid interface stricter than necessary to prevent errors
   // dispatch valid implies dispatch ready
   assert(!a_int_dispatch.valid || a_int_dispatch.ready)
   assert(!b_int_dispatch.valid || b_int_dispatch.ready)
   assert(!a_mem_dispatch.valid || a_mem_dispatch.ready)
   assert(!b_mem_dispatch.valid || b_mem_dispatch.ready)
+  if(usingFPU) {
+    assert(!a_fp_dispatch.get.valid || a_fp_dispatch.get.ready)
+  }
 
   // dispatch implies dequeue
   assert(!a_int_dispatch.valid || a_queue.io.deq_uop)
   assert(!b_int_dispatch.valid || b_queue.io.deq_uop)
   assert(!a_mem_dispatch.valid || a_queue.io.deq_uop)
   assert(!b_mem_dispatch.valid || b_queue.io.deq_uop)
+  if(usingFPU) {
+    assert(!a_fp_dispatch.get.valid || a_queue.io.deq_uop)
+  }
 
   // dequeue implies dispatch
-  assert(!a_queue.io.deq_uop || (a_int_dispatch.valid || a_mem_dispatch.valid))
+  assert(!a_queue.io.deq_uop || (a_int_dispatch.valid || a_mem_dispatch.valid || a_fp_dispatch.map(_.valid).getOrElse(false.B)))
```

Comment:
Why  a_fp_dispatch.map(_.valid).getOrElse(false.B)) instead og a_fp_dispatch.valid?

Reply:
because a_fp_dispatch is an Option - basically a fancy way of only using it if it is not None

Reply:
a_fp_dispatch is a port option
a_fp_dispatch.map(_.valid) is a Bool option
.getOrElse(false.B)) takes the Bool if it exists (Option is not None) and false.B otherwise

Reply:
You can think of options as lists with wither 1 or 0 elements - map gives you a new list and .getOrElse either the element or an alternative